src/main/java/bio/terra/workspace/service/resource/controlled/ControlledResourceService.java

```diff
@@ -31,17 +36,59 @@
   private final WorkspaceService workspaceService;
   private final ResourceDao resourceDao;
   private final StageService stageService;
+  private final SamService samService;
 
   @Autowired
   public ControlledResourceService(
       JobService jobService,
       WorkspaceService workspaceService,
       ResourceDao resourceDao,
-      StageService stageService) {
+      StageService stageService,
+      SamService samService) {
     this.jobService = jobService;
     this.workspaceService = workspaceService;
     this.resourceDao = resourceDao;
     this.stageService = stageService;
+    this.samService = samService;
+  }
+
+  /**
+   * Convenience function that checks existence of a controlled resource within a workspace,
+   * followed by an authorization check against that resource.
+   *
+   * <p>Throws ResourceNotFound from getResource if the workspace does not exist in the specified
+   * workspace, regardless of the user's permission.
+   *
+   * <p>Throws InvalidControlledResourceException if the given resource is not controlled.
+   *
+   * <p>Throws SamUnauthorizedException if the user is not permitted to perform the specified action
+   * on the resource in question.
+   *
+   * <p>Returns the controlled resource object if it exists and the user is permitted to perform the
+   * specified action.
```

Comment:
I appreciate the comment. Very clear

Reply:
credit to @davidangb, this is "heavily inspired" by `validateWorkspaceAndAction`